phastSim/phastSim.py

```diff
@@ -641,23 +809,196 @@ def sampleMutation(self, allele, rates, rand):
+            deleted_leaves_left = 0
+            if rand > rate0 or remaining_deletions == 0:
+                # under these conditions we can short circuit and ignore this whole branch
+                pass
+
+            else:
+                if child0.level < level or not (child0.isTerminal and child0.allele == "-"):
+                    newChild0 = genomeNode(level=level)
+                    newChild0.isTerminal = child0.isTerminal
+                    newChild0.rate = child0.rate
+
+                    if child0.isTerminal:
+                        newChild0.refNode = child0.refNode
+                        newChild0.allele = child0.allele
+                    else:
+                        newChild0.belowNodes = list(child0.belowNodes)
+
+                
+                deleted_leaves_left = self.deleteNodes(rand, newChild0, remaining_deletions, level)
+                newRate0 = newChild0.rate
+
+            # deal with right child but with slightly different exit conditions
+            deleted_leaves_right = 0
+            if rand > rate0 + rate1 or remaining_deletions - deleted_leaves_left == 0:
+                # can short circuit this child under slightly different conditions
+                pass
+
+            else:
+                if child1.level < level or not (child1.isTerminal and child1.allele == "-"):
+                    newChild1 = genomeNode(level=level)
+                    newChild1.isTerminal = child1.isTerminal
+                    newChild1.rate = child1.rate
+
+                    if child1.isTerminal:
+                        newChild1.refNode = child1.refNode
+                        newChild1.allele = child1.allele
+
+                    else:
+                        newChild1.belowNodes = list(child1.belowNodes)
+       
+
+                deleted_leaves_right = self.deleteNodes(rand - rate0, newChild1, remaining_deletions - deleted_leaves_left, level)
+                newRate1 = newChild1.rate
+
+            node.rate = newRate0 + newRate1
+            node.belowNodes = [newChild0, newChild1]
+
+            return deleted_leaves_left + deleted_leaves_right
+
+    def buildTree(self, node, insert_list, pos_left, pos_right):
+
+        if pos_left == pos_right:
+            # then we are at a leaf, we need to build a new leaf node from scratch
+            node.isTerminal = True
+            node.refNode = node
+            if self.codon:
+                node.allele = self.codonAlleles[insert_list[0]]
+            else:
+                
+                node.allele = self.alleles[insert_list[0]]
+                node.rates = self.mutMatrixRepeats[-1]
+
+            node.genomePos = node.genomePos = [(self.n_insertions, pos_left), (self.n_insertions, pos_right)]
+            node.insertionRate = self.insertionRate[-1]
+            node.deletionRate = self.deletionRate[-1]
+            node.deletionLength = self.deletionLength[-1]
+            node.insertionLength = self.insertionLength[-1]
```

Comment:
Ok for now - as these are constant at this stage, but in future these will need to be sampled from prior distributions, the same that is done for nodes in the initial genome tree.